### PR TITLE
Address tokens request being retried

### PR DESCRIPTION
**Pull request by desiprisg**
## Type of change

- [x] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [x] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

Only retry getToken requests with an exponential backoff if the status is not 4XX. If it is, remove the session so the automatic polling stops.

### CODE REVIEW

File: packages/shared/src/utils/errors.ts [hunk 1]
```diff
@@ -4,6 +4,11 @@ export function isUnauthorizedError(e: any): boolean {
   return code === 'authentication_invalid' && status === 401;
 }
 
+export function is400Error(e: any): boolean {
```
Comment (SokratisVidros) [inline packages/shared/src/utils/errors.ts:7]:
```suggestion
export function is4xxError(e: any): boolean {
```

File: .changeset/sixty-yaks-stare.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (SokratisVidros) [inline .changeset/sixty-yaks-stare.md:1]: ❓ Is this file required?

**Review by SokratisVidros — APPROVED** [review]

File: .changeset/sixty-yaks-stare.md (hunk 2 above)
Comment (panteliselef) [reply .changeset/sixty-yaks-stare.md:1]: Yes, a changeset file needs to be committed, but we can leave it empty

**Review by anagstef — APPROVED** [review]

File: packages/shared/src/utils/errors.ts (hunk 1 above)
Comment (desiprisg) [reply packages/shared/src/utils/errors.ts:7]: I did it in the commit messages and not here smh

File: packages/clerk-js/src/core/resources/Session.ts [hunk 3]
```diff
@@ -66,7 +67,7 @@ export class Session extends BaseResource implements SessionResource {
 
   getToken: GetToken = async (options?: GetTokenOptions): Promise<string | null> => {
     return runWithExponentialBackOff(() => this._getToken(options), {
-      shouldRetry: (e: unknown, i: number) => !isUnauthorizedError(e) || i < 8,
+      shouldRetry: (e: unknown, i: number) => !is4xxError(e) && i < 4,
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/core/resources/Session.ts:70]:
```suggestion
      shouldRetry: (error: unknown, currentIteration: number) => !is4xxError(error) && currentIteration < 4,
```

File: packages/clerk-js/src/core/services/authentication/SessionCookieService.ts [hunk 4]
```diff
@@ -64,7 +64,11 @@ export class SessionCookieService {
     try {
       this.updateSessionCookie(await this.getNewToken());
     } catch (e) {
-      return this.handleGetTokenError(e);
+      try {
+        return this.handleGetTokenError(e);
+      } catch {
+        return this.clerk.setActive({ session: null });
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/core/services/authentication/SessionCookieService.ts:70]: If I'm not mistaken, there's already logic that handles the unauthenticated state here: 
https://github.com/clerkinc/javascript/blob/3b21b60091afe52a3276621c48142facb04c62d0/packages/clerk-js/src/core/resources/Base.ts#L66

Do we need to add more error codes to the above line? 
Can we avoid the need of this nested try/catch clause?

Comment (desiprisg) [reply packages/clerk-js/src/core/services/authentication/SessionCookieService.ts]: Yeah, works well. Can't get rid of the try catch block though.

File: packages/clerk-js/src/core/services/authentication/SessionCookieService.ts [hunk 5]
```diff
@@ -64,7 +64,11 @@ export class SessionCookieService {
     try {
       this.updateSessionCookie(await this.getNewToken());
     } catch (e) {
-      return this.handleGetTokenError(e);
+      try {
+        return this.handleGetTokenError(e);
+      } catch {
+        void this.clerk.handleUnauthenticated();
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/core/services/authentication/SessionCookieService.ts:70]: Can we move the nested catch inside `handleGetTokenError`?

**Review by nikosdouvlis — APPROVED** [review]
Good job

File: packages/shared/src/utils/errors.ts [hunk 6]
```diff
@@ -4,6 +4,11 @@ export function isUnauthorizedError(e: any): boolean {
   return code === 'authentication_invalid' && status === 401;
 }
 
+export function is4xxError(e: any): boolean {
+  const status = e?.status;
+  return status >= 400 && status < 500;
```
Comment (nikosdouvlis) [inline packages/shared/src/utils/errors.ts:9]:
```suggestion
  return !!status &&  status >= 400 && status < 500;
```

File: packages/clerk-js/src/core/services/authentication/SessionCookieService.ts [hunk 7]
```diff
@@ -96,12 +97,19 @@ export class SessionCookieService {
   }
 
   private handleGetTokenError(e: any) {
+    if (is4xxError(e)) {
+      void this.clerk.handleUnauthenticated();
+      return;
+    }
+
+    if (isNetworkError(e)) {
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/core/services/authentication/SessionCookieService.ts:105]: Can you please add a few comments here? line 100 + line 105

File: packages/clerk-js/src/core/services/authentication/SessionCookieService.ts (hunk 5 above)
Comment (desiprisg) [reply packages/clerk-js/src/core/services/authentication/SessionCookieService.ts]: done

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
